### DISCUSSION TITLE
Spotify broken for USB DAC

### DISCUSSION THREAD

**Discussion by Roland-B** [rern/rAudio#1213] [Bug Reports]
Latest image, latest update applied. Spotify fails to connect. 

Roland.

Output of Spotify renderer:

device = "hw:0"
control = "hw:0"
mixer = "Headphone"
device = "hw:1"
control = "hw:1"
mixer = "E30 "
device = "hw:1"
control = "hw:1"
mixer = "E30 "
device = "hw:1"
control = "hw:1"
mixer = "E30 "
device = "hw:0"
control = "hw:0"
mixer = "Headphone"
device = "hw:0"
control = "hw:0"
mixer = "E30 "
device = "hw:0"
control = "hw:0"
mixer = "E30 "
device = "hw:0"
control = "hw:0"
mixer = "E30 "
device = "hw:0"
control = "hw:0"
mixer = "E30 "

spotifyd.service- A spotify playing daemon
     Loaded: loaded (/etc/systemd/system/spotifyd.service; enabled; preset: disabled)
     Active: active (running) since Sun 2023-01-29 14:54:27 GMT; 13s ago
   Main PID: 10102 (spotifyd)
      Tasks: 2 (limit: 4524)
        CPU: 20ms
     CGroup: /system.slice/spotifyd.service
             └─10102 /usr/bin/spotifyd --no-daemon

Jan 29 14:54:27 rAudio systemd[1]: Started A spotify playing daemon.
Jan 29 14:54:27 rAudio spotifyd[10102]: Loading config from "/etc/spotifyd.conf"
Jan 29 14:54:27 rAudio spotifyd[10102]: No username specified. Checking username_cmd
Jan 29 14:54:27 rAudio spotifyd[10102]: No username_cmd specified
Jan 29 14:54:27 rAudio spotifyd[10102]: No password specified. Checking password_cmd
Jan 29 14:54:27 rAudio spotifyd[10102]: No password_cmd specified
Jan 29 14:54:27 rAudio spotifyd[10102]: No proxy specified
Jan 29 14:54:27 rAudio spotifyd[10102]: Using software volume controller.
Jan 29 14:54:27 rAudio spotifyd[10102]: no usable credentials found, enabling discovery
Jan 29 14:54:27 rAudio spotifyd[10102]: Failed to register IPv6 receiver: Os { code: 97, kind: Uncategorized, message: "Address family not supported by protocol" }

**Comment by R101G** [comment]
@Roland-B 

I asked a while ago if it was possible for any new features to be in the form of addons, rather than part of the core player, but I know nothing about writing code, so that might not be practical. Obviously, this is rern's project, shared for free, and I don't think rern was the one who chose Arch Linux as a base.

As for testing, I have no idea how many rAudio users there are, but I would be willing to bet that you would not find two of them who have exactly the same setup and use rAudio in the same way. All I need is a system which will perform the functions of a radio tuner and a CD transport, and rAudio is brilliant at this. I would not notice any problems with features which are of no interest me, and I tolerate the temporary regressions rather than use an old version because I like (most of) the improvements to the basic features and functions. There are times when a bug-fix-only LTS version would seem to be a good idea, though!

**Reply by digixmax** [reply]
> ...
> All I need is a system which will perform the functions of a radio tuner and a CD transport, and rAudio is brilliant at this.
> ...

Sorry for being off topic, but may I ask which kind of CD/optical drive do you use with your RPi for this CD transport system?

TIA.

**Reply by R101G** [reply]
@digixmax 
I mean that rAudio eliminates my need for a tuner and CD transport. Apparently, you can still plug in a USB DAB adapter and optical drive if you want to.

**Comment by HunLucas** [comment]
RPI 4 8GB, with dac+ pro and without. same problem R101G helped me to turn ipv6 on, but unfortunately doesnt fix it

UPDATE
Works again with today's update...20230130 Thanks Rern

**Comment by Roland-B** [comment]
@rern. Still not fixed for me. The build keeps changing the config to a different one that breaks Spotify for me.

The build keeps changing my "spotifyd.conf" to the following ->

[global]
bitrate = 320
onevent = "/srv/http/bash/spotifyd.sh"
use_mpris = false
backend = "alsa"
volume_controller = "alsa"
**device = "hw:0"**
control = "hw:0"
mixer = "E30 "

The correct settings to work for me is

[global]
bitrate = 320
onevent = "/srv/http/bash/spotifyd.sh"
use_mpris = false
backend = "alsa"
volume_controller = "alsa"
**device = "default:CARD=E30"**
control = "hw:0"
mixer = "E30 "


**Edit : Checked this and doesnt seem to be the cause but am leaving here for info**: - I believe this is may be due to the fact I have power saving enabled on the E30 DAC so when not in use it powers down but then one of your scripts then rewrites the spotifyd.conf file to change the device setting when I go to use Spotify again. 

The following file below and contents is what I have always used until recently without any issue on runeaudio and rAudio1 (not sure which build it broke with as only recently was forced to update to keep getting updates). 

/etc/udev/rules.d/usb-power.rules >

ACTION=="add", SUBSYSTEM=="usb", ATTR{idVendor}=="152a", ATTR{idProduct}=="8750", TEST=="power/control", ATTR{power/control}="auto"

Would be greatful if you could help with this. Thanks.

Roland B.

**Comment by AndyFx12** [comment]
@Roland-B 
> @rern Can we just have ONE update that doesnt break something when fixing something else please its getting rather frustrating now!!! There should be 2 builds if your not willing to test your builds fully before releasing.. One standard build that stays as is with any fixes needed and tested properly and one for new features which can then be tested by the people who request them. Always seems to be one step forward, 2 steps back recently. Thanks.

Hi Roland, 
I don´t agree! rAudio has so much features - it´s nearly impossible to test them "fully". Rern is permanently working on refactoring his code, to avoid side-effects. But nobody can test all combinations of hardware / settings. And nobody is forced to update - you can read release notes at addons-site. If you don´t need the update - don´t press the button ...

@rern - maybe it´s possible to get an option to disable update-checks? 

regards, Andy

**Reply by Roland-B** [reply]
@AndyFx12. I don't care if you agree or not but thanks for bumping the thread. Request a feature change correctly elsewhere though not on a bug report. Thanks.

Roland.

**Comment by Roland-B** [comment]
@rern - Latest image (rAudio-64bit-20230212), no settings imported. Only thing configured is on board audio disabled, and Spotify enabled.

Spotifyd@rAudio is shown as available in Spotify to connect to, when selected it it connects briefly with no sound then stops and crashes Spotifyd

Output of spotify renderer is : -

spotifyd 0.3.4-1
# cat /etc/spotifyd.conf
[global]
bitrate = 320
onevent = "/srv/http/bash/spotifyd.sh"
use_mpris = false
backend = "alsa"
volume_controller = "alsa"
device = "hw:0"
control = "hw:0"
mixer = "E30 "

spotifyd.service- A spotify playing daemon
     Loaded: loaded (/etc/systemd/system/spotifyd.service; enabled; preset: disabled)
     Active: active (running) since Mon 2023-02-13 18:25:01 UTC; 19s ago
   Main PID: 3474 (spotifyd)
      Tasks: 4 (limit: 4521)
        CPU: 5.219s
     CGroup: /system.slice/spotifyd.service
             ├─3474 /usr/bin/spotifyd --no-daemon
             ├─3895 /bin/bash /srv/http/bash/spotifyd.sh
             └─3907 "[spotifyd.sh]"

Feb 13 18:25:19 rAudio spotifyd[3474]: Alsa error PCM open ALSA function 'snd_pcm_hw_params_set_format' failed with error 'EINVAL: Invalid argument'
Feb 13 18:25:19 rAudio spotifyd[3474]: Could not start audio: Alsa error: PCM open failed
Feb 13 18:25:19 rAudio spotifyd[3474]: The application panicked (crashed).
Feb 13 18:25:19 rAudio spotifyd[3474]: Message:  called `Option::unwrap()` on a `None` value
Feb 13 18:25:19 rAudio spotifyd[3474]: Location: /build/.cargo/registry/src/github.com-1ecc6299db9ec823/librespot-playback-0.2.0/src/audio_backend/alsa.rs:160
Feb 13 18:25:19 rAudio spotifyd[3474]: Backtrace omitted.
Feb 13 18:25:19 rAudio spotifyd[3474]: Run with RUST_BACKTRACE=1 environment variable to display it.
Feb 13 18:25:19 rAudio spotifyd[3474]: Run with RUST_BACKTRACE=full to include source snippets.
Feb 13 18:25:19 rAudio spotifyd[3474]: Running "/srv/http/bash/spotifyd.sh" using "/bin/bash" with environment variables {"TRACK_ID": "3mpPFykAKO6DTXUapHsoFm", "PLAY_REQUEST_ID": "0", "DURATION_MS": "226926", "PLAYER_EVENT": "play", "POSITION_MS": "197718"}
Feb 13 18:25:21 rAudio spotifyd[3474]: Running "/srv/http/bash/spotifyd.sh" using "/bin/bash" with environment variables {"PLAYER_EVENT": "play", "DURATION_MS": "226926", "PLAY_REQUEST_ID": "0", "POSITION_MS": "197722", "TRACK_ID": "3mpPFykAKO6DTXUapHsoFm"}


Journalctl Output : -

Feb 13 18:43:04 rAudio spotifyd[1715]: Connecting to AP "ap.spotify.com:443"
Feb 13 18:43:05 rAudio spotifyd[1715]: Authenticated as "**REMOVEDFORPRIVACY**" !
Feb 13 18:43:05 rAudio spotifyd[1715]: Country: "DE"
Feb 13 18:43:05 rAudio spotifyd[1715]: Using Alsa sink with format: S16
Feb 13 18:43:05 rAudio spotifyd[1715]: Running "/srv/http/bash/spotifyd.sh" using "/bin/bash" with environment variables {"VOLUME": "6>
Feb 13 18:43:05 rAudio spotifyd[1715]: Loading <Big Blue> with Spotify URI <spotify:track:3mpPFykAKO6DTXUapHsoFm>
Feb 13 18:43:05 rAudio spotifyd[1715]: Running "/srv/http/bash/spotifyd.sh" using "/bin/bash" with environment variables {"PLAY_REQUES>
Feb 13 18:43:06 rAudio spotifyd[1715]: <Big Blue> (226926 ms) loaded
Feb 13 18:43:06 rAudio spotifyd[1715]: Running "/srv/http/bash/spotifyd.sh" using "/bin/bash" with environment variables {"PLAYER_EVEN>
Feb 13 18:43:06 rAudio spotifyd[1715]: Alsa error PCM open ALSA function 'snd_pcm_hw_params_set_format' failed with error 'EINVAL: Inv>
Feb 13 18:43:06 rAudio spotifyd[1715]: Could not start audio: Alsa error: PCM open failed
Feb 13 18:43:06 rAudio spotifyd[1715]: Alsa error PCM open ALSA function 'snd_pcm_hw_params_set_format' failed with error 'EINVAL: Inv>
Feb 13 18:43:06 rAudio spotifyd[1715]: Could not start audio: Alsa error: PCM open failed
Feb 13 18:43:06 rAudio spotifyd[1715]: The application panicked (crashed).
Feb 13 18:43:06 rAudio spotifyd[1715]: Message:  called `Option::unwrap()` on a `None` value
Feb 13 18:43:06 rAudio spotifyd[1715]: Location: /build/.cargo/registry/src/github.com-1ecc6299db9ec823/librespot-playback-0.2.0/src/a>
Feb 13 18:43:06 rAudio spotifyd[1715]: Backtrace omitted.
Feb 13 18:43:06 rAudio spotifyd[1715]: Run with RUST_BACKTRACE=1 environment variable to display it.
Feb 13 18:43:06 rAudio spotifyd[1715]: Run with RUST_BACKTRACE=full to include source snippets.
Feb 13 18:43:07 rAudio spotifyd[1715]: Running "/srv/http/bash/spotifyd.sh" using "/bin/bash" with environment variables {"DURATION_MS>
Feb 13 18:43:09 rAudio spotifyd[1715]: Running "/srv/http/bash/spotifyd.sh" using "/bin/bash" with environment variables {"POSITION_MS>
Feb 13 18:43:11 rAudio spotifyd[1715]: Running "/srv/http/bash/spotifyd.sh" using "/bin/bash" with environment variables {"TRACK_ID": >
Feb 13 18:43:13 rAudio spotifyd[1715]: The application panicked (crashed).
Feb 13 18:43:13 rAudio spotifyd[1715]: Message:  called `Option::unwrap()` on a `None` value
Feb 13 18:43:13 rAudio spotifyd[1715]: Location: src/main_loop.rs:194
Feb 13 18:43:13 rAudio spotifyd[1715]: Backtrace omitted.
Feb 13 18:43:13 rAudio spotifyd[1715]: Run with RUST_BACKTRACE=1 environment variable to display it.
Feb 13 18:43:13 rAudio spotifyd[1715]: Run with RUST_BACKTRACE=full to include source snippets.
Feb 13 18:43:13 rAudio spotifyd[1715]: Player thread panicked!
Feb 13 18:43:13 rAudio systemd[1]: spotifyd.service: Main process exited, code=exited, status=101/n/a
Feb 13 18:43:13 rAudio audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=spotifyd comm="systemd" exe="/usr/l>
Feb 13 18:43:13 rAudio systemd[1]: spotifyd.service: Failed with result 'exit-code'.




Please can you look to fix as does not appear to be detecting the USB DAC correctly which when working used to be (**device = "default:CARD=E30"**)

Roland B.

rAudio 20230212
6.1.11-1 aarch64
Raspberry Pi 4B Rev 1.2
BCM2711 • 3.7 GiB
4 x Cortex-A72 @ 1.50 GHz

**Comment by TinkeringB** [comment]
All was fine using the headphones jack, but I had a DAC lying around and thought to use that.
Connected IFI iDAC usb -> no sound in Spotify since, while internet radio is working fine.

Found a solution 

rAudio 20230310
5.15.78-1 armv7l
Raspberry Pi 2B Rev 1.1
BCM2836 • 959 MiB
4 x Cortex-A7 @ 900 MHz

Connected IFI iDAC usb -> no sound in Spotify, while internet radio is working fine.

Got error : "Alsa error PCM open ALSA function 'snd_pcm_hw_params_set_format' failed with error 'EINVAL: Invalid argument'"


found: https://github.com/dtcooper/raspotify/issues/235

changed  /etc/spotifyd.conf
Original not working /etc/spotifyd.conf:
```
[global]
bitrate = 320
onevent = "/srv/http/bash/spotifyd.sh"
use_mpris = false
backend = "alsa"
volume_controller = "alsa"
device = "hw:2"
control = "hw:2"
mixer = "AMR Audio 2.0 Output"
```
Working /etc/spotifyd.conf after restarting the service
```
[global]
bitrate = 320
onevent = "/srv/http/bash/spotifyd.sh"
use_mpris = false
backend = "alsa"
volume_controller = "alsa"
device = "plughw:2"
control = "hw:2"
mixer = "AMR Audio 2.0 Output"
```

note the change for 'device = "plughw:2"

after restarting the spodyfy service, I got sound :-)
```
systemctl restart spotifyd
```

**Comment by TinkeringB** [comment]
After having issues with this setup not being able to controll the volume anymore; (possibly after installing the DAP plugin??) I am setting up a new instance using an RPI3.
Again, I am running into this no audio issue for my DAC. Apparently the above method (plughw) fails this time.

But, found a better solution I think. first I ran aplay -L, and determined the correct device, in this case "default:CARD=A20"
```
pi3raudio:~  #  aplay -L
null
    Discard all samples (playback) or generate zero samples (capture)
default
    Default Audio Device
sysdefault
    Default Audio Device
default:CARD=A20
    AMR USB Audio 2.0, USB Audio
    Default Audio Device
sysdefault:CARD=A20
    AMR USB Audio 2.0, USB Audio
    Default Audio Device
front:CARD=A20,DEV=0
    AMR USB Audio 2.0, USB Audio
    Front output / input
surround21:CARD=A20,DEV=0
    AMR USB Audio 2.0, USB Audio
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=A20,DEV=0
    AMR USB Audio 2.0, USB Audio
    4.0 Surround output to Front and Rear speakers
surround41:CARD=A20,DEV=0
    AMR USB Audio 2.0, USB Audio
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=A20,DEV=0
    AMR USB Audio 2.0, USB Audio
    5.0 Surround output to Front, Center and Rear speakers
surround51:CARD=A20,DEV=0
    AMR USB Audio 2.0, USB Audio
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=A20,DEV=0
    AMR USB Audio 2.0, USB Audio
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
iec958:CARD=A20,DEV=0
    AMR USB Audio 2.0, USB Audio
    IEC958 (S/PDIF) Digital Audio Output
pi3raudio:~  #
```

Then Edited  /etc/spotifyd.conf and replaced original "hw:0,0" with "default:CARD=A20" found above
```
[global]
bitrate = 320
onevent = "/srv/http/bash/spotifyd.sh"
use_mpris = false
backend = "alsa"
volume_controller = "alsa"
device = "default:CARD=A20"
control = "hw:0,0"
mixer = "AMR Audio 2.0 Output"
```
and restarted  spotifyd.

--> audio :-)


P.S.
Changes made using the cog-wheel behind Spotity are not being applied nor saved somehow; 
![image](https://github.com/rern/rAudio/assets/9662310/e458a339-ac1d-4aa7-a1fc-81bfb8ecd616)



P.S.2 
Sadly my changes are reverted after boot and 'device' jumped back to non-working "hw:0,0"

**Reply by daoane** [reply]
Have a look at https://github.com/rern/rAudio/discussions/1118
